compiler/rustc_codegen_llvm/src/back/archive.rs

```diff
@@ -332,22 +406,53 @@ impl<'a> LlvmArchiveBuilder<'a> {
         }
     }
 
-    fn i686_decorated_name(import: &DllImport) -> CString {
+    fn i686_decorated_name(import: &DllImport, mingw: bool) -> String {
         let name = import.name;
-        // We verified during construction that `name` does not contain any NULL characters, so the
-        // conversion to CString is guaranteed to succeed.
-        CString::new(match import.calling_convention {
-            DllCallingConvention::C => format!("_{}", name),
-            DllCallingConvention::Stdcall(arg_list_size) => format!("_{}@{}", name, arg_list_size),
+        let prefix = if mingw { "" } else { "_" };
+
+        match import.calling_convention {
+            DllCallingConvention::C => format!("{}{}", prefix, name),
+            DllCallingConvention::Stdcall(arg_list_size) => {
+                format!("{}{}@{}", prefix, name, arg_list_size)
+            }
             DllCallingConvention::Fastcall(arg_list_size) => format!("@{}@{}", name, arg_list_size),
             DllCallingConvention::Vectorcall(arg_list_size) => {
                 format!("{}@@{}", name, arg_list_size)
             }
-        })
-        .unwrap()
+        }
     }
 }
 
 fn string_to_io_error(s: String) -> io::Error {
     io::Error::new(io::ErrorKind::Other, format!("bad archive: {}", s))
 }
+
+fn find_binutils_dlltool(sess: &Session) -> OsString {
+    assert!(sess.target.options.is_like_windows && !sess.target.options.is_like_msvc);
+    if let Some(dlltool_path) = &sess.opts.debugging_opts.dlltool {
+        return dlltool_path.clone().into_os_string();
+    }
+    if sess.host.arch != sess.target.arch {
+        let prefixed_dlltool = if sess.target.arch == "x86" {
+            "i686-w64-mingw32-dlltool"
+        } else {
+            "x86_64-w64-mingw32-dlltool"
+        };
+        let prefixed_dlltool = if cfg!(windows) {
+            [prefixed_dlltool, "exe"].concat()
+        } else {
+            prefixed_dlltool.to_string()
+        };
+        for dir in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
+            let full_path = dir.join(&prefixed_dlltool);
+            if full_path.is_file() {
+                return full_path.into_os_string();
+            }
+        }
+    }
+    // REVIEWER FEEDBACK REQUESTED: a few lines above, we look for $PREFIXED_DLLTOOL.exe or
+    // $PREFIXED_DLLTOOL in the path, depending on whether or not rustc is running on Windows.
+    // Do we need to do the same thing here?  And why do we need to search PATH explicitly
+    // above, but not here?
+    OsString::from("dlltool.exe")
```

Comment:
Yes, it might be cleaner to determine the name of the tool once at the beginning, something like:
```rust
fn find_binutils_dlltool(sess: &Session) -> OsString {
    assert!(sess.target.options.is_like_windows && !sess.target.options.is_like_msvc);
    if let Some(dlltool_path) = &sess.opts.debugging_opts.dlltool {
        return dlltool_path.clone().into_os_string();
    }
    let mut tool_name: OsString = if sess.host.arch != sess.target.arch {
        // We are cross compiling, so we need the tool with the appropriate prefix
        if sess.target.arch == "x86" {
            "i686-w64-mingw32-dlltool"
        } else {
            "x86_64-w64-mingw32-dlltool"
        }
    } else {
        // We are not cross compiling, so we just want `dlltool`
        "dlltool"
    }.into();

    if sess.host.options.is_like_windows {
        // If we are compiling on Windows, we also need the .exe suffix
        tool_name.push(".exe")
    }

    // NOTE: I'm unsure how useful it is to explicitly search PATH.
    for dir in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
        let full_path = dir.join(&tool_name);
        if full_path.is_file() {
            return full_path.into_os_string();
        }
    }

    // The appropriate dlltool does not seem to be available on PATH. 
    // Just return the name of the tool and let the invocation fail with 
    // with a hopefully useful error message.
    tool_name
}
```